Awkward 2.5.2 causing test failure

### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of anndata.
- [X] (optional) I have confirmed this bug exists on the master branch of anndata.

### Report

`test_awkward_io` is currently failing with awkward 2.5.2. The issue has to do with us passing in a numpy integer, while the code is currently expecting a python integer. I think this is an awkward array bug since this didn't error in awkward 2.5.1.

I propose we just pin against awkward 2.5.2.

It's unclear if this would cause problems in real world usage. It's interesting that only one of our cases is failing, and that it's an empty array. I may be okay with unpinning it once a fix is out.

Reported upstream at:

* https://github.com/scikit-hep/awkward/issues/2947

```sh
pytest -k test_awkward_io
```

<details>
<summary> Test failure</summary>

```pytb
================================================================================== test session starts ==================================================================================
platform darwin -- Python 3.11.6, pytest-7.4.4, pluggy-1.3.0
rootdir: /Users/isaac/github/anndata
configfile: pyproject.toml
testpaths: anndata, docs/concatenation.rst
plugins: cov-4.1.0, mock-3.11.1, memray-1.5.0, anyio-4.1.0, profiling-1.7.0, xdist-3.3.1, nunit-1.0.4
collected 3039 items / 3026 deselected / 13 selected                                                                                                                                    

anndata/tests/test_awkward.py .......F.....                                                                                                                                       [100%]

======================================================================================= FAILURES ========================================================================================
________________________________________________________________________________ test_awkward_io[array7] ________________________________________________________________________________

tmp_path = PosixPath('/private/var/folders/bd/43q20k0n6z15tdfzxvd22r7c0000gn/T/pytest-of-isaac/pytest-18/test_awkward_io_array7_0')
array = <Array [[], [], [], [], [], [], []] type='7 * 0 * float64'>

    @pytest.mark.parametrize(
        "array",
        [
            # numpy array
            ak.Array(np.arange(2 * 3 * 4 * 5).reshape((2, 3, 4, 5))),
            # record
            ak.Array([{"a": 1, "b": 2}, {"a": 1, "b": 3}]),
            # ListType, variable length
            ak.Array([[1], [2, 3], [4, 5, 6]]),
            # RegularType + nested ListType
            ak.to_regular(ak.Array([[[1, 2], [3]], [[2], [3, 4, 5]]]), 1),
            # nested record
            ak.to_regular(ak.Array([[{"a": 0}, {"b": 1}], [{"c": 2}, {"d": 3}]]), 1),
            # mixed types (variable length)
            ak.Array([[1, 2], ["a"]]),
            # zero-size edge cases
            ak.Array(np.ones((0, 7))),
            ak.Array(np.ones((7, 0))),
            # UnionType of two regular types with different dimensions
            ak.concatenate([ak.Array(np.ones((2, 2))), ak.Array(np.ones((2, 3)))]),
            # UnionType of two regular types with same dimension
            ak.concatenate(
                [
                    ak.Array(np.ones((2, 2))),
                    ak.Array(np.array([["a", "a"], ["a", "a"]])),
                ]
            ),
            # categorical array
            ak.str.to_categorical(ak.Array([["a", "b", "c"], ["a", "b"]])),
            ak.str.to_categorical(ak.Array([[1, 1, 2], [3, 3]])),
            # tyical record type with AIRR data consisting of different dtypes
            ak.Array(
                [
                    [
                        {
                            "v_call": "TRV1",
                            "junction_aa": "ADDEEKK",
                            "productive": True,
                            "locus": None,
                            "consensus_count": 3,
                        },
                        {
                            "v_call": "TRV2",
                            "productive": False,
                            "locus": "TRA",
                            "consensus_count": 4,
                        },
                    ],
                    [
                        {
                            "v_call": None,
                            "junction_aa": "ADDEKK",
                            "productive": None,
                            "locus": "IGK",
                            "consensus_count": 3,
                        }
                    ],
                ]
            ),
        ],
    )
    def test_awkward_io(tmp_path, array):
        adata = AnnData()
        adata.uns["awk"] = array
        adata_path = tmp_path / "adata.h5ad"
        adata.write_h5ad(adata_path)
    
        adata2 = read_h5ad(adata_path)
    
>       adata2.uns["awk"].to_list()

anndata/tests/test_awkward.py:246: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniforge3/envs/scanpy-dev/lib/python3.11/site-packages/awkward/highlevel.py:496: in to_list
    return self._layout.to_list(self._behavior)
../../miniforge3/envs/scanpy-dev/lib/python3.11/site-packages/awkward/contents/content.py:1086: in to_list
    return self.to_packed()._to_list(behavior, None)
../../miniforge3/envs/scanpy-dev/lib/python3.11/site-packages/awkward/contents/regulararray.py:1464: in to_packed
    content = self._content[: index_nplike.shape_item_as_index(length)]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <awkward._nplikes.numpy.Numpy object at 0x128af7d10>, x1 = 0

    def shape_item_as_index(self, x1: ShapeItem) -> int:
        if x1 is unknown_length:
            raise TypeError("array module nplikes do not support unknown lengths")
        elif isinstance(x1, int):
            return x1
        else:
>           raise TypeError(f"expected None or int type, received {x1}")
E           TypeError: expected None or int type, received 0

../../miniforge3/envs/scanpy-dev/lib/python3.11/site-packages/awkward/_nplikes/array_module.py:323: TypeError
=================================================================================== warnings summary ====================================================================================
anndata/tests/test_awkward.py: 26 warnings
  /Users/isaac/github/anndata/anndata/_core/anndata.py:183: ImplicitModificationWarning: Transforming to str index.
    warnings.warn("Transforming to str index.", ImplicitModificationWarning)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================ short test summary info ================================================================================
FAILED anndata/tests/test_awkward.py::test_awkward_io[array7] - TypeError: expected None or int type, received 0
============================================================== 1 failed, 12 passed, 3026 deselected, 26 warnings in 1.82s ===============================================================
```

</details>

### Versions

```python
-----
IPython             8.17.2
anndata             0.11.0.dev46+g0219fefb.d20240115
awkward             2.5.2
numpy               1.23.4
session_info        1.0.0
-----
asciitree           NA
asttokens           NA
attr                23.1.0
awkward_cpp         NA
cloudpickle         3.0.0
cython_runtime      NA
dask                2023.10.0
dateutil            2.8.2
decorator           5.1.1
executing           2.0.1
fasteners           0.19
h5py                3.10.0
importlib_metadata  NA
jedi                0.19.1
jinja2              3.1.2
markupsafe          2.1.3
msgpack             1.0.7
natsort             8.4.0
numcodecs           0.12.1
packaging           23.2
pandas              2.1.1
parso               0.8.3
pexpect             4.8.0
prompt_toolkit      3.0.41
psutil              5.9.6
ptyprocess          0.7.0
pure_eval           0.2.2
pyarrow             14.0.1
pygments            2.16.1
pytz                2023.3.post1
rich                NA
scipy               1.11.4
setuptools          68.2.2
setuptools_scm      NA
six                 1.16.0
sphinxcontrib       NA
stack_data          0.6.3
tblib               3.0.0
tlz                 0.12.0
toolz               0.12.0
traitlets           5.13.0
wcwidth             0.2.10
yaml                6.0.1
zarr                2.16.1
zipp                NA
-----
Python 3.11.6 | packaged by conda-forge | (main, Oct  3 2023, 10:40:37) [Clang 15.0.7 ]
macOS-13.6.1-x86_64-i386-64bit
-----
Session information updated at 2024-01-15 13:46
```

--- a/anndata/_io/specs/methods.py
+++ b/anndata/_io/specs/methods.py
@@ -645,7 +645,7 @@ def read_awkward(elem, _reader):
     length = _read_attr(elem.attrs, "length")
     container = {k: _reader.read_elem(elem[k]) for k in elem.keys()}
 
-    return ak.from_buffers(form, length, container)
+    return ak.from_buffers(form, int(length), container)
 
 
 ##############

--- a/docs/release-notes/0.10.5.md
+++ b/docs/release-notes/0.10.5.md
@@ -5,6 +5,7 @@
 
 * Fix outer concatenation along variables when only a subset of objects had an entry in layers {pr}`1291` {user}`ivirshup`
 * Fix comparison of >2d arrays in `uns` during concatenation {pr}`1300` {user}`ivirshup`
+* Fix IO with awkward array version 2.5.2 {pr}`1328` {user}`ivirshup`
 
 ```{rubric} Documentation
 ```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -92,7 +92,7 @@ test = [
     "scanpy",
     "httpx", # For data downloading
     "dask[array,distributed]",
-    "awkward>=2.3, !=2.5.2",
+    "awkward>=2.3",
     "pyarrow",
     "pytest_memray",
 ]

